feat: add dependency approach doc

matthew-chirgwin opened the pull request:
- adds section to cover dependency approach

Closes: #31

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
feat: add dependency approach doc

- adds section to cover dependency approach

Contributes to: #31

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

JQrdan commented on docs/Architecture.md:
```diff
@@ -278,6 +279,23 @@ For this capability to work, [file name conventions](#file-name-conventions) nee
 
+#### Dependency management
+
+This UI will make use of a number of third party dependencies to operate at develop, test, and runtime. The will range from helper tools such as Webpack, to frameworks, such as React. The `package.json` file/schema defines three types of dependency, `dependencies`, `peerDependencies` and `devDependencies`. Depending on which section a dependency sits, it will be treated differently at build time, ...
```
The second sentence reads weird to me

JQrdan commented on docs/Architecture.md:
```diff
@@ -278,6 +279,23 @@ For this capability to work, [file name conventions](#file-name-conventions) nee
+
+The Strimzi-ui will not be published. Instead, it will consume a number of packages, and be built to provide a UI. Some of these packages will be used and shipped directly, but some will also generate and polyfill code in this repository to produce the the built UI. Thus, in the event of a security issue, we need to understand if the issue relates in any way to what we ship. We therefore categori ...
+
+- `dependencies` are any dependency which are either shipped bundled in the built output via direct usage (eg React), included in the dockerfile which will run the UI (eg Express) or generate output that is then built and shipped (ie babel transforms/plugins, Webpack plugins etc)
+- `devDependencies` are used to support the development and test of the UI, but that are not shipped. Eg Webpack, Jest, eslint
+
```
E.g. and I.e. instead of eg and ie

JQrdan reviewed (changes requested):
Sounds good to me, just some minor changes needed

matthew-chirgwin committed:
refactor: wording

- updates wording as per review comments

Contributes to: #32

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

JQrdan reviewed (approved):
LGTM

dlabaj commented on docs/Architecture.md:
```diff
@@ -167,9 +168,9 @@ Further details will be added to this section once https://github.com/strimzi/pr
 #### Entity model
 
-A UI typically allows a user to interact with resources. These resources are types of known entity (eg a Topic), which typically can be created, read, updated or deleted. Similarly, entities are often protected by authorization controls, restricting a user's access or ability to create, read, update or delete a given named entity or entity type. In the case of authorization for example, different ...
+A UI typically allows a user to interact with resources. These resources are types of known entity (E.g. a Topic), which typically can be created, read, updated or deleted. Similarly, entities are often protected by authorization controls, restricting a user's access or ability to create, read, update or delete a given named entity or entity type. In the case of authorization for example, differe ...
```
I think we can get rid of the word Similarly and just use "Entities are often..."

dlabaj replied:
Same with Equally... could just say "The ability..."

matthew-chirgwin replied:
Both updated - good suggestions :)

dlabaj commented on docs/Architecture.md:
```diff
@@ -278,6 +279,23 @@ For this capability to work, [file name conventions](#file-name-conventions) nee
 
+#### Dependency management
+
+This UI will make use of a number of third party dependencies to operate at develop, test, and runtime. This will range from helper tools such as Webpack, to frameworks, like React. The `package.json` file/schema defines three types of dependency, `dependencies`, `peerDependencies` and `devDependencies`. Depending on which section a dependency sits, it will be treated differently at `npm publish` ...
```
We could get rid of the `npm publish` part since we are not publishing this package, and change it to `npm build` everywhere instead.

matthew-chirgwin replied:
This was intended to be a quick summary of the normal npm approach to dependencies, with the following sentences/paragraphs explaining the difference in approach we take. I will update the wording to be more clear here.

dlabaj replied:
No worries.  I just didn't want someone to get confused reading this.  Thanks again.

dlabaj commented on docs/Architecture.md:
```diff
@@ -278,6 +279,23 @@ For this capability to work, [file name conventions](#file-name-conventions) nee
+
+Tooling, such as [`npm audit`](https://docs.npmjs.com/cli/audit) make use of/utilise these sections to report when a dependency has a known vulnerability, and can thus call out if an issue is shipped by this package, assuming that dependencies are correctly catagorised based on their usage. Given the nature of JavaScript and where it is used, having these tools report issues when they occur is in ...
+
+The Strimzi-ui will not be published. Instead, it will consume a number of packages, and be built to provide a UI. Some of these packages will be used and shipped directly, but some will also generate and polyfill code in this repository to produce the the built UI. Thus, in the event of a security issue, we need to understand if the issue relates in any way to what we ship. We therefore categori ...
```
Can we consolidate this section with the one above, so we only have 3 bullets instead of 5 bullets (dependencies, devDependencies, and peerDependencies)?

dlabaj commented on docs/Architecture.md:
```diff
@@ -278,6 +279,23 @@ For this capability to work, [file name conventions](#file-name-conventions) nee
+
+The Strimzi-ui will not be published. Instead, it will consume a number of packages, and be built to provide a UI. Some of these packages will be used and shipped directly, but some will also generate and polyfill code in this repository to produce the the built UI. Thus, in the event of a security issue, we need to understand if the issue relates in any way to what we ship. We therefore categori ...
+
+- `dependencies` are any dependency which are either shipped bundled in the built output via direct usage (E.g. React), included in the dockerfile which will run the UI (E.g. Express) or generate output that is then built and shipped (I.e. babel transforms/plugins, Webpack plugins etc)
```
I noticed `npm audit` does audit devDependencies as well.  Wondering if we can get rid of the generate output part of this.  I believe having them in devDependencies will work correctly for us.

matthew-chirgwin replied:
`npm audit` is one tool which can list both normal and dev dependencies. There are others which cannot/do not however (just reading the dependencies section). Plus, some generated code is included in/amongst the code we write in the built package, as if we imported it.

dlabaj replied:
okay then this is fine.  Thanks for the clarification.

dlabaj reviewed (commented):
Thanks for submitting this.  A few comments / questions.

matthew-chirgwin commented:
Thank you for the comments @dlabaj / @JQrdan - I have rebased and updated the PR given these. If there are no more comments I will look to merge this tomorrow
